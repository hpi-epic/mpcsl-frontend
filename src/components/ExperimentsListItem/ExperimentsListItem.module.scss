.ListItem {
  max-width: 300px;
  min-width: 300px;
  word-wrap: break-word;
  margin: 5px;
  flex-grow: 1;
}

.ListItemContent {
  height: 200px;
  display: flex;
  flex-direction: column;
  p {
    flex-grow: 1;
  }
}

.DropdownButton {
  border: none;
  background-color: transparent;
  &hover {
    border: none;
    background-color: transparent;
  }
}

.ListButton {
  margin-right: 5px;
}

.Description {
  width: 100%;
}
