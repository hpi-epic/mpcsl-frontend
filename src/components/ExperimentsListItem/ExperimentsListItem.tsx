import React, { useState, useEffect } from 'react';
import {
  Select,
  Modal,
  Dropdown,
  Menu,
  Button,
  Card,
  Badge,
  Icon,
  Tooltip,
  InputNumber,
  Form,
  Descriptions
} from 'antd';
import { IExperiment, BadgeStatus, IAlgorithm } from '../../types';
import { useHistory } from 'react-router-dom';
import {
  deleteExperiment,
  getK8SNodes,
  runExperiment,
  getAlgorithm,
  getJobsForExperiment
} from '../../actions/apiRequests';
import styles from './ExperimentsListItem.module.scss';
import { FormComponentProps } from 'antd/lib/form';
const { Option } = Select;
const { confirm } = Modal;

const ExperimentDropdown = (
  props: IExperiment & { onView: () => void; onDuplicate: () => void }
) => {
  const history = useHistory();
  return (
    <Dropdown
      overlay={
        <Menu>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                history.push(
                  `/${props.dataset_id}/experiments/${props.id}/jobs`
                );
              }}
              type="primary"
              ghost={true}
            >
              View Jobs
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                props.onView();
              }}
              key="1"
            >
              View Settings
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                props.onDuplicate();
              }}
              key="3"
            >
              Duplicate
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                confirm({
                  title: 'Do you want to delete the following Experiment?',
                  content: `${props.name} - ${props.description}`,
                  onOk() {
                    deleteExperiment(props).catch();
                  }
                });
              }}
              type="danger"
              ghost={true}
              key="4"
            >
              Delete
            </Button>
          </Menu.Item>
        </Menu>
      }
      placement="bottomLeft"
    >
      <Icon type="ellipsis" onClick={e => e.stopPropagation()} />
    </Dropdown>
  );
};

type RunExperimentModalProps = FormComponentProps & {
  experiment: IExperiment;
  visible: boolean;
  onClose: () => void;
};

const RunExperimentModal: React.FunctionComponent<RunExperimentModalProps> = props => {
  const { getFieldDecorator } = props.form;
  const [k8sNodes, setK8sNodes] = useState<undefined | string[]>();
  const [needsGPU, setNeedsGPU] = useState<undefined | boolean>();
  useEffect(() => {
    getK8SNodes()
      .then(setK8sNodes)
      .catch(() => setK8sNodes([]));
  }, []);
  useEffect(() => {
    getAlgorithm(props.experiment.algorithm_id).then(alg =>
      setNeedsGPU(!!alg.needs_gpu)
    );
  }, [props.experiment]);
  return (
    <Modal
      title="Select Machine to Start Job"
      visible={props.visible}
      confirmLoading={needsGPU === undefined}
      onOk={() => {
        props.form.validateFields(
          (
            err: Error,
            values: {
              node: string;
              runs: number;
              parallelismMode: 'parallel' | 'sequential';
              gpus?: number;
            }
          ) => {
            if (!err) {
              runExperiment(
                props.experiment,
                values.node === '_none' ? undefined : values.node,
                values.runs,
                values.parallelismMode === 'parallel',
                needsGPU ? values.gpus : undefined
              );
              props.onClose();
            }
          }
        );
      }}
      onCancel={() => props.onClose()}
    >
      <Form layout="vertical">
        <Form.Item label="Set Execution Parameters">
          {getFieldDecorator('node', { initialValue: '_none' })(
            <Select>
              <Option value="_none" style={{ fontStyle: 'italic' }}>
                Default
              </Option>
              {k8sNodes
                ? k8sNodes.map(node => (
                    <Option key={node} value={node}>
                      {node}
                    </Option>
                  ))
                : null}
            </Select>
          )}
        </Form.Item>
        <Form.Item label="Number of Jobs">
          {getFieldDecorator('runs', { initialValue: 1 })(
            <InputNumber min={1} style={{ width: '100%' }} />
          )}
        </Form.Item>
        <Form.Item label="Run mode">
          {getFieldDecorator('parallelismMode', { initialValue: 'parallel' })(
            <Select>
              <Option value="parallel">Parallel</Option>
              <Option value="sequential">Sequential</Option>
            </Select>
          )}
        </Form.Item>
        {needsGPU ? (
          <Form.Item label="GPU Count">
            {getFieldDecorator('gpus', { initialValue: 1 })(
              <InputNumber min={1} style={{ width: '100%' }} />
            )}
          </Form.Item>
        ) : null}
      </Form>
    </Modal>
  );
};

const RunExperimentModalForm = Form.create<RunExperimentModalProps>()(
  RunExperimentModal
);

const ExperimentsListItem = (
  props: IExperiment & {
    onView: (experimentId: number) => void;
    onDuplicate: (experimentId: number) => void;
  }
) => {
  const [nodeSelectModal, setNodeSelectModal] = useState(false);
  const [algorithm, setAlgorithm] = useState<undefined | IAlgorithm>();
  const [jobCount, setJobCount] = useState<undefined | number>();
  const history = useHistory();
  useEffect(() => {
    getAlgorithm(props.algorithm_id).then(setAlgorithm);
  }, [props.algorithm_id]);
  useEffect(() => {
    getJobsForExperiment(props.id).then(jobs => setJobCount(jobs.length));
  }, [props.id]);
  const { name, description, last_job, execution_time_statistics } = props;
  const statusText = last_job ? last_job.status : 'not started';
  return (
    <>
      <RunExperimentModalForm
        onClose={() => setNodeSelectModal(false)}
        visible={nodeSelectModal}
        experiment={props}
      />
      <Card
        title={
          <div style={{ display: 'flex', flexDirection: 'column', height: 50 }}>
            <div style={{ display: 'flex', justifyContent: 'space-between' }}>
              <p>{name}</p>
              <Badge
                className="Card-Badge"
                text={statusText}
                status={
                  last_job?.status ? BadgeStatus[last_job.status] : 'default'
                }
              />
            </div>
            {execution_time_statistics ? (
              <i style={{ fontWeight: 'lighter', fontSize: '0.8rem' }}>
                Mean Execution Time: {execution_time_statistics.mean.toFixed(3)}
                s
              </i>
            ) : null}
          </div>
        }
        actions={[
          <Tooltip key="run" title="Run Experiment">
            <Icon
              style={{ fontSize: 20 }}
              type="play-circle"
              onClick={e => {
                e.stopPropagation();
                setNodeSelectModal(true);
              }}
            />
          </Tooltip>,
          <Tooltip key="compare" title="Compare Experiment">
            <Icon
              style={{ fontSize: 20 }}
              type="interaction"
              onClick={e => {
                e.stopPropagation();
                if (last_job) {
                  history.push(
                    `/${props.dataset_id}/experiments/${props.id}/compare`
                  );
                }
              }}
            />
          </Tooltip>,
          <ExperimentDropdown
            key="dropdown"
            {...props}
            onView={() => props.onView(props.id)}
            onDuplicate={() => props.onDuplicate(props.id)}
          />
        ]}
        hoverable
        className={styles.ListItem}
        onClick={() => {
          if (props.last_job && props.last_job?.result) {
            history.push(
              `/${props.dataset_id}/experiments/${props.id}/jobs/${props.last_job?.result?.id}`
            );
          }
        }}
      >
        <div className={styles.ListItemContent}>
          <Descriptions size="small" column={2}>
            <Descriptions.Item
              className={styles.Description}
              span={2}
              label="Description"
            >
              <p
                style={{
                  height: 84,
                  overflow: 'hidden',
                  width: 250,
                  overflowWrap: 'break-word'
                }}
              >
                {description}
              </p>
            </Descriptions.Item>
            {algorithm ? (
              <>
                <Descriptions.Item label="Package">
                  {algorithm.package}
                </Descriptions.Item>
                <Descriptions.Item label="Function">
                  {algorithm.function}
                </Descriptions.Item>
              </>
            ) : null}
            {props.last_job ? (
              <Descriptions.Item span={2} label="Last Job run at">
                {new Date(props.last_job.start_time).toLocaleString()}
              </Descriptions.Item>
            ) : null}
            {props.last_job ? (
              <Descriptions.Item label="Job Count">
                {jobCount ? jobCount : 'Loading...'}
              </Descriptions.Item>
            ) : null}
          </Descriptions>
        </div>
      </Card>
    </>
  );
};

export { ExperimentsListItem };
