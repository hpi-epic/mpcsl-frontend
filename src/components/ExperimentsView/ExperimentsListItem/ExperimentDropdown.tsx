import { Button, Dropdown, Icon, Menu, Modal } from 'antd';
import React from 'react';
import { useHistory } from 'react-router-dom';
import { deleteExperiment } from '../../../restAPI/apiRequests';
import { IExperiment } from '../../../types/types';
import styles from './ExperimentsListItem.module.scss';
const { confirm } = Modal;

export const ExperimentDropdown = (props: {
  experiment: IExperiment;
  onView: () => void;
  onDuplicate: () => void;
}) => {
  const history = useHistory();
  return (
    <Dropdown
      overlay={
        <Menu>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                history.push(
                  `/${props.experiment.dataset_id}/experiments/${props.experiment.id}/jobs`
                );
              }}
              type="primary"
              ghost={true}
            >
              View Jobs
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                props.onView();
              }}
              key="1"
            >
              View Settings
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                props.onDuplicate();
              }}
              key="3"
            >
              Duplicate
            </Button>
          </Menu.Item>
          <Menu.Item>
            <Button
              className={styles.DropdownButton}
              onClick={e => {
                e.stopPropagation();
                confirm({
                  title: 'Do you want to delete the following Experiment?',
                  content: `${props.experiment.name} - ${props.experiment.description}`,
                  onOk() {
                    deleteExperiment(props.experiment).catch();
                  }
                });
              }}
              type="danger"
              ghost={true}
              key="4"
            >
              Delete
            </Button>
          </Menu.Item>
        </Menu>
      }
      placement="bottomLeft"
    >
      <Icon type="ellipsis" onClick={e => e.stopPropagation()} />
    </Dropdown>
  );
};
