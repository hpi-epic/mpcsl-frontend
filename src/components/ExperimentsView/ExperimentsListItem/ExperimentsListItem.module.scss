.ListItem {
  max-width: 300px;
  min-width: 300px;
  word-wrap: break-word;
  margin: 5px;
  flex-grow: 1;
}

.ListItemContent {
  height: 200px;
  display: flex;
  flex-direction: column;

  p {
    flex-grow: 1;
  }
}

.DropdownButton {
  border: none;
  background-color: transparent !important;
  box-shadow: none;

  &hover {
    border: none;
    background-color: transparent;
  }
}

.ListButton {
  margin-right: 5px;
}

.Description {
  color: rgba(0, 0, 0, 0.65);
  width: 100%;
}
