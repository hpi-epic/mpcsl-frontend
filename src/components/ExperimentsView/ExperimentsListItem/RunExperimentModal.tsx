import { Form, InputNumber, Modal, Select } from 'antd';
import React, { useEffect, useState } from 'react';
import {
  getAlgorithm,
  getK8SNodes,
  runExperiment
} from '../../../restAPI/apiRequests';
import { IExperiment } from '../../../types/types';

const { Option } = Select;

type IProps = {
  experiment: IExperiment;
  visible: boolean;
  onClose: () => void;
};

export const RunExperimentModalForm: React.FunctionComponent<IProps> = props => {
  const [form] = Form.useForm();

  const [k8sNodes, setK8sNodes] = useState<undefined | string[]>();
  const [needsGPU, setNeedsGPU] = useState<undefined | boolean>();

  useEffect(() => {
    getK8SNodes()
      .then(setK8sNodes)
      .catch(() => setK8sNodes([]));
  }, []);

  useEffect(() => {
    getAlgorithm(props.experiment.algorithm_id).then(alg =>
      setNeedsGPU(!!alg.needs_gpu)
    );
  }, [props.experiment]);

  return (
    <Modal
      title="Set Execution Parameters"
      visible={props.visible}
      confirmLoading={needsGPU === undefined}
      onOk={() => {
        form
          .validateFields()
          .then(values => {
            runExperiment(
              props.experiment,
              values.node === '_none' ? undefined : values.node,
              values.runs,
              values.parallelismMode === 'parallel',
              needsGPU ? values.gpus : undefined,
              values.enforceCpus
            );
            props.onClose();
          })
          .catch(info => {
            console.log('Validation failed:', info);
          });
      }}
      onCancel={() => props.onClose()}
    >
      <Form form={form} layout="vertical">
        <Form.Item
          label="Select Machine to Start Job"
          name="node"
          initialValue={'_none'}
        >
          <Select>
            <Option value="_none" style={{ fontStyle: 'italic' }}>
              Default
            </Option>
            {k8sNodes
              ? k8sNodes.map(node => (
                  <Option key={node} value={node}>
                    {node}
                  </Option>
                ))
              : null}
          </Select>
        </Form.Item>
        <Form.Item label="Number of Jobs" name="runs" initialValue={1}>
          <InputNumber min={1} style={{ width: '100%' }} />
        </Form.Item>
        <Form.Item
          label="Run Mode"
          name="parallelismMode"
          initialValue={'parallel'}
        >
          <Select>
            <Option value="parallel">Parallel</Option>
            <Option value="sequential">Sequential</Option>
          </Select>
        </Form.Item>
        <Form.Item
          label="Enforce CPU Requests"
          name="enforceCpus"
          initialValue={1}
        >
          <Select>
            <Option value={1}>Yes</Option>
            <Option value={0}>No</Option>
          </Select>
        </Form.Item>
        {needsGPU ? (
          <Form.Item label="GPU Count" name="gpus" initialValue={1}>
            <InputNumber min={1} style={{ width: '100%' }} />
          </Form.Item>
        ) : null}
      </Form>
    </Modal>
  );
};
