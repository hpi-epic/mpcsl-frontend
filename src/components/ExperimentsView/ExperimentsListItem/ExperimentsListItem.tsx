import { Badge, Card, Descriptions, Icon, Tooltip } from 'antd';
import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import {
  getAlgorithm,
  getJobsForExperiment
} from '../../../restAPI/apiRequests';
import { BadgeStatus, IAlgorithm, IExperiment } from '../../../types/types';
import { ExperimentDropdown } from './ExperimentDropdown';
import styles from './ExperimentsListItem.module.scss';
import { RunExperimentModalForm } from './RunExperimentModal';

export const ExperimentsListItem = (props: {
  experiment: IExperiment;
  onView: (experimentId: number) => void;
  onDuplicate: (experimentId: number) => void;
}) => {
  const {
    id,
    algorithm_id,
    name,
    description,
    last_job,
    execution_time_statistics
  } = props.experiment;
  const [nodeSelectModal, setNodeSelectModal] = useState(false);
  const [algorithm, setAlgorithm] = useState<undefined | IAlgorithm>();
  const [jobCount, setJobCount] = useState<undefined | number>();
  const history = useHistory();
  useEffect(() => {
    getAlgorithm(algorithm_id).then(setAlgorithm);
  }, [algorithm_id]);
  useEffect(() => {
    getJobsForExperiment(id).then(jobs => setJobCount(jobs.length));
  }, [id, last_job]);
  const statusText = last_job ? last_job.status : 'not started';
  return (
    <>
      <RunExperimentModalForm
        onClose={() => setNodeSelectModal(false)}
        visible={nodeSelectModal}
        experiment={props.experiment}
      />
      <Card
        title={
          <div style={{ display: 'flex', flexDirection: 'column', height: 50 }}>
            <div style={{ display: 'flex', justifyContent: 'space-between' }}>
              <p>{name}</p>
              <Badge
                className="Card-Badge"
                text={statusText}
                status={
                  last_job?.status ? BadgeStatus[last_job.status] : 'default'
                }
              />
            </div>
            {execution_time_statistics ? (
              <i style={{ fontWeight: 'lighter', fontSize: '0.8rem' }}>
                Mean Execution Time: {execution_time_statistics.mean.toFixed(3)}
                s
              </i>
            ) : null}
          </div>
        }
        actions={[
          <Tooltip key="run" title="Run Experiment">
            <Icon
              style={{ fontSize: 20 }}
              type="play-circle"
              onClick={e => {
                e.stopPropagation();
                setNodeSelectModal(true);
              }}
            />
          </Tooltip>,
          <Tooltip key="compare" title="Compare Experiment">
            <Icon
              style={{ fontSize: 20 }}
              type="interaction"
              onClick={e => {
                e.stopPropagation();
                if (last_job) {
                  history.push(
                    `/${props.experiment.dataset_id}/experiments/${props.experiment.id}/compare`
                  );
                }
              }}
            />
          </Tooltip>,
          <ExperimentDropdown
            key="dropdown"
            experiment={props.experiment}
            onView={() => props.onView(props.experiment.id)}
            onDuplicate={() => props.onDuplicate(props.experiment.id)}
          />
        ]}
        hoverable
        className={styles.ListItem}
        onClick={() => {
          if (last_job && last_job?.result) {
            history.push(
              `/${props.experiment.dataset_id}/experiments/${props.experiment.id}/jobs/${last_job?.result?.id}`
            );
          }
        }}
      >
        <div className={styles.ListItemContent}>
          <Descriptions size="small" column={1}>
            <Descriptions.Item
              className={styles.Description}
              label="Description"
            >
              <p
                style={{
                  height: 58,
                  overflow: 'hidden',
                  width: 250,
                  overflowWrap: 'break-word'
                }}
              >
                {description}
              </p>
            </Descriptions.Item>
            {algorithm ? (
              <>
                <Descriptions.Item label="Package">
                  {algorithm.package}
                </Descriptions.Item>
                <Descriptions.Item label="Function">
                  {algorithm.function}
                </Descriptions.Item>
              </>
            ) : null}
            {last_job ? (
              <Descriptions.Item label="Last Job run">
                {new Date(last_job.start_time).toLocaleString('de-DE')}
              </Descriptions.Item>
            ) : null}
            {last_job ? (
              <Descriptions.Item label="Job Count">
                {jobCount ? jobCount : 'Loading...'}
              </Descriptions.Item>
            ) : null}
          </Descriptions>
        </div>
      </Card>
    </>
  );
};
