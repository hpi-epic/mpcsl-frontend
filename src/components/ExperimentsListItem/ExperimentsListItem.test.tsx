import React from 'react';
import renderer from 'react-test-renderer';
import { ExperimentsListItem } from './ExperimentsListItem';
import { MemoryRouter } from 'react-router-dom';

describe('<ListElementExperiment />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <ExperimentsListItem
            dataset_id={0}
            id={0}
            name="test"
            algorithm_id={0}
            parameters={{}}
            onDuplicate={() => undefined}
            onView={() => undefined}
          />
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
