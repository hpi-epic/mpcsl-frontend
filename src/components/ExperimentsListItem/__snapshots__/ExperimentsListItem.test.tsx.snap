// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListElementExperiment /> renders correctly 1`] = `
<div
  className="ant-card ListItem ant-card-bordered ant-card-hoverable"
  onClick={[Function]}
>
  <div
    className="ant-card-head"
    style={Object {}}
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        <div
          style={
            Object {
              "display": "flex",
              "flexDirection": "column",
              "height": 50,
            }
          }
        >
          <div
            style={
              Object {
                "display": "flex",
                "justifyContent": "space-between",
              }
            }
          >
            <p>
              test
            </p>
            <span
              className="Card-Badge ant-badge ant-badge-status ant-badge-not-a-wrapper"
            >
              <span
                className="ant-badge-status-dot ant-badge-status-default"
                style={Object {}}
              />
              <span
                className="ant-badge-status-text"
                style={
                  Object {
                    "color": undefined,
                  }
                }
              >
                not started
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="ListItemContent"
    >
      <p />
    </div>
  </div>
  <ul
    className="ant-card-actions"
  >
    <li
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <span>
        <i
          aria-label="icon: play-circle"
          className="anticon anticon-play-circle"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="play-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M719.4 499.1l-296.1-215A15.9 15.9 0 0 0 398 297v430c0 13.1 14.8 20.5 25.3 12.9l296.1-215a15.9 15.9 0 0 0 0-25.8zm-257.6 134V390.9L628.5 512 461.8 633.1z"
            />
          </svg>
        </i>
      </span>
    </li>
    <li
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <span>
        <i
          aria-label="icon: interaction"
          className="anticon anticon-interaction"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="interaction"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656zM304.8 524h50.7c3.7 0 6.8-3 6.8-6.8v-78.9c0-19.7 15.9-35.6 35.5-35.6h205.7v53.4c0 5.7 6.5 8.8 10.9 5.3l109.1-85.7c3.5-2.7 3.5-8 0-10.7l-109.1-85.7c-4.4-3.5-10.9-.3-10.9 5.3V338H397.7c-55.1 0-99.7 44.8-99.7 100.1V517c0 4 3 7 6.8 7zm-4.2 134.9l109.1 85.7c4.4 3.5 10.9.3 10.9-5.3v-53.4h205.7c55.1 0 99.7-44.8 99.7-100.1v-78.9c0-3.7-3-6.8-6.8-6.8h-50.7c-3.7 0-6.8 3-6.8 6.8v78.9c0 19.7-15.9 35.6-35.5 35.6H420.6V568c0-5.7-6.5-8.8-10.9-5.3l-109.1 85.7c-3.5 2.5-3.5 7.8 0 10.5z"
            />
          </svg>
        </i>
      </span>
    </li>
    <li
      style={
        Object {
          "width": "33.333333333333336%",
        }
      }
    >
      <span>
        <i
          aria-label="icon: ellipsis"
          className="anticon anticon-ellipsis ant-dropdown-trigger"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          tabIndex={-1}
        >
          <svg
            aria-hidden="true"
            className=""
            data-icon="ellipsis"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M176 511a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm280 0a56 56 0 1 0 112 0 56 56 0 1 0-112 0zm280 0a56 56 0 1 0 112 0 56 56 0 1 0-112 0z"
            />
          </svg>
        </i>
      </span>
    </li>
  </ul>
</div>
`;
